pkg/infra/block_collector.go

```diff
@@ -0,0 +1,39 @@
+package infra
+
+import "sync"
+
+type BlockCollector struct {
+	sync.Mutex
+	threshold int
+	registry  map[uint64]int
+}
+
+func NewBlockCollector(threshold int) *BlockCollector {
+	registry := make(map[uint64]int)
+	return &BlockCollector{
+		threshold: threshold,
+		registry:  registry,
+	}
+}
+
+func (bc *BlockCollector) Commit(block uint64) (committed bool) {
+	if bc.threshold == 1 {
+		return true
+	}
+	bc.Lock()
+	defer bc.Unlock()
+	// read map
+	remaining, ok := bc.registry[block]
+	if !ok {
+		bc.registry[block] = bc.threshold - 1
+		return false
+	}
+
+	if remaining == 1 {
+		delete(bc.registry, block)
+		return true
```

Comment:
when peer1 and peer2 completed the process.
here comes peer3. which never reach here and memory leak?